 Copyright © Sorbonne University.

 This source code is licensed under the MIT license found in the
 LICENSE file in the root directory of this source tree.

## Binômes :
- ***Abdelkrim HALIMI***
- ***Hatem AYED***

# Outlook

In this notebook, you will code a naive actor-critic algorithm in the tabular case. Then you will tune it using grid search and Bayesian optimization, potentially using the [optuna](https://optuna.readthedocs.io/en/stable/) library.
Finally, you will get the best hyper-parameters obtained with both methods and perform a statistical test to see if there is a statistically significant difference between these methods and with respect to naive hyper-parameter values.

## Install libraries

In [1]:
# Installs the necessary Python and system libraries
try:
    from easypip import easyimport, easyinstall, is_notebook
except ModuleNotFoundError as e:
    get_ipython().run_line_magic("pip", "install 'easypip>=1.2.0'")
    from easypip import easyimport, easyinstall, is_notebook

easyinstall("swig")
easyinstall("bbrl>=0.2.2")
easyinstall("bbrl_gymnasium>=0.2.0")
easyinstall("tensorboard")
easyinstall("moviepy")
easyinstall("box2d-kengz")
easyinstall("optuna")
easyinstall("gymnasium")
easyinstall("mazemdp")

import numpy as np
import os
from typing import List

import hydra
import optuna
import yaml
from omegaconf import OmegaConf, DictConfig

# For visualization
os.environ["VIDEO_FPS"] = "5"
if not os.path.isdir("./videos"):
    os.mkdir("./videos")

from IPython.display import Video

In [2]:
import torch
import torch.nn as nn
import numpy as np 
import seaborn as sns 

In [3]:
import gymnasium as gym

from bbrl.utils.chrono import Chrono

import matplotlib
import matplotlib.pyplot as plt

from mazemdp.toolbox import sample_categorical
from mazemdp.mdp import Mdp
from bbrl_gymnasium.envs.maze_mdp import MazeMDPEnv
from gymnasium.wrappers.monitoring.video_recorder import VideoRecorder
from functools import partial

matplotlib.use("TkAgg")

Matplotlib backend: module://matplotlib_inline.backend_inline


# Step 1: Coding the naive Actor-critic algorithm

We consider the naive actor-critic algorithm with a categorical policy.
The algorithm learns a critic with the standard temporal difference mechanism
using a learning rate $\alpha_{critic}$.

We consider a value-based critic $V(s)$. The extension to an action value function $Q(s,a)$ is straightforward.

To update the critic, the algorithm computes the temporal difference error:

$$\delta_t = r(s_t, a_t) + \gamma V^{(n)}(s_{t+1})-V^{(n)}(s_t).$$

Then it applies it to the critic:

$$V^{(n+1)}(s_t) = V^{(n)}(s_t) + \alpha_{critic} \delta_t.$$

To update the actor, the general idea is the same, using the temporal difference error with another learning rate $\alpha_{actor}$.

However, naively applying the same learning rule would not ensure that the probabilities of all actions in a state sum to 1.
Besides, when the temporal difference error $\delta_t$ is negative, it may happen that the probability of an action gets negative or null, which raises an issue when applying renormalization.

So, instead of applying the naive rule, we apply the following one:
$$ 
\pi_{temp}(a_t|s_t) =  \begin{cases}
\pi^{(i)}(a_t|s_t) + \alpha_{actor} \delta_t & \mathrm{if } \pi^{(i)}(a_t|s_t) + \alpha_{actor} \delta_t > 10^{-8}\\
10^{-8} & \mathrm{otherwise.} \\
\end{cases}
$$

Then we can apply renormalization so that the probabilities of actions still sum to 1, with
$$
\forall a, \pi^{(i+1)}(a|s_t) = \frac{\pi_{temp}^{(i+1)}(a|s_t)} {\sum_{a'} \pi_{temp}^{(i+1)}(a'|s_t)}
$$ with
$$ 
\pi_{temp}^{(i+1)}(a|s_t) =  \begin{cases}
\pi_{temp}(a|s_t) & \mathrm{if } a = a_t\\
\pi^{(i)}(a|s_t) & \mathrm{otherwise.} \\
\end{cases}
$$

## Exercise 1

### 1. Code the naive actor-critic algorithm as specified above.

Some hints:

- a good idea to build this code it to take inspiration from the code of Q-learning, to add an actor (a categorical policy), both learning rates,
and to take care about the renormalization function.

- for the next steps of this lab, having a function to repeatedly call your actor-critic algorithm and save the learning trajectories and
norms of the value function is a good idea.

In [4]:
import gymnasium as gym
import numpy as np 

# Environment with 20% of walls and no negative reward when hitting a wall
env = gym.make(
    "MazeMDP-v0",
    kwargs={"width": 4, "height": 3, "ratio": 0.2, "hit": 0.0},
    render_mode="human",
)
env.reset()
env.unwrapped.init_draw("The maze")

       [ 1,  4,  5,  8],
       [ 2, -1,  6,  9]]), t…

In [5]:
def get_policy_from_actor(actor: np.ndarray) -> np.ndarray:
    return np.argmax(actor,axis=1)

def actor_critic(
    mdp: MazeMDPEnv,
    nb_episodes: int = 20,
    timeout: int = 50,
    alpha_critic : float = 0.5,
    alpha_actor : float=0.5,
    render: bool = True,
) -> tuple[np.ndarray, List[float]]:
    """
    Args:
        mdp: the environment
        nb_episodes: the number of episodes
        timeout: the maximum number of steps per episode
        alpha_critic: the critic learning rate
        alpha_actor: the actor learning rate
        render: whether to render the environment
    Returns:
        a_probs: the learned policy
        v: the learned value function
        v_list: the norm of the value function at
        time_list: the number of steps per episode
        mean_reward: the mean reward
    """

    # Initialize the state-action value function
    # alpha is the learning rate
    v = np.zeros(mdp.unwrapped.nb_states)  # initial state value v
    a_probs = np.ones((mdp.unwrapped.nb_states, mdp.action_space.n))/mdp.action_space.n 
    v_min = np.zeros(mdp.unwrapped.nb_states)
    v_list = []
    time_list = []
    reward_list = []
    # Run learning cycle
    mdp.timeout = timeout  # episode length

    if render:
        mdp.init_draw("Actor Critic")

    for _ in range(nb_episodes):
        
        reward = 0
        # Draw the first state of episode i using a uniform distribution over all the states
        x, _ = mdp.reset(uniform=True)
        cpt = 0

        terminated = False
        truncated = False

        while not (terminated or truncated):
            if render:
                mdp.draw_v_pi(v, get_policy_from_actor(a_probs))

            # Draw an action using a soft-max policy
            u = sample_categorical(a_probs[x])
            y, r, terminated, truncated, _ = mdp.step(u)
            # Update the state-action value function with q-Learning
            delta = r + mdp.gamma*v[y]*(1-terminated) - v[x]
            v[x] = v[x] + alpha_critic*delta

            temp = a_probs[x,u] + alpha_actor*delta
            a_probs[x, u] = temp if temp > 1e-8 else 1e-8
            a_probs[x] = a_probs[x]/a_probs[x].sum()

            x = y
            cpt = cpt + 1
            reward += r

        reward_list.append(reward)
        v_list.append(np.linalg.norm(np.maximum(v, v_min)))
        time_list.append(cpt)

    if render:
        mdp.current_state = 0
        mdp.draw_v_pi(v, get_policy_from_actor(a_probs))

    return a_probs, v, v_list, time_list, reward_list

In [6]:
NB_EPISODES = 100
TIMEOUT = 50
ALPHA_CRITIC = 0.5
ALPHA_ACTOR = 0.5

a, v, v_list, time_list, reward_list = actor_critic(
    env, alpha_critic=ALPHA_CRITIC,alpha_actor=ALPHA_ACTOR, nb_episodes=NB_EPISODES, timeout=TIMEOUT
)

/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/gymnasium/core.py:311: UserWarning: WARN: env.init_draw to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.init_draw` for environment variables or `env.get_wrapper_attr('init_draw')` that will search the reminding wrappers.
  logger.warn(


       [ 1,  4,  5,  8],
       [ 2, -1,  6,  9]]), t…

/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/gymnasium/core.py:311: UserWarning: WARN: env.draw_v_pi to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.draw_v_pi` for environment variables or `env.get_wrapper_attr('draw_v_pi')` that will search the reminding wrappers.
  logger.warn(
/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/gymnasium/core.py:311: UserWarning: WARN: env.gamma to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.gamma` for environment variables or `env.get_wrapper_attr('gamma')` that will search the reminding wrappers.
  logger.warn(


In [7]:
plt.figure(figsize=(12, 7))
sns.heatmap(a, annot=True, fmt=".2f", cmap="viridis")
plt.title("Actor probabilities")
plt.show()

### 2. Provide a plot function

Your plot function should show the evolution through time of number of steps the agent takes to find the reward in the maze.
If your algorithm works, this number of steps should decrease through time.

Your plot function should also show a mean and a standard deviation (or some more advanced statistics) over a collection of learning runs.

In [8]:
class Plotter():

    @staticmethod
    def moving_average(data, window_size):
        return np.convolve(data, np.ones(window_size)/window_size, mode='valid')

    @staticmethod
    def plot_steps_evolution(time_list, save_path="figure1"):
        mean = np.round(np.mean(time_list), 3)
        std = np.round(np.std(time_list), 3)
        q1 = np.round(np.percentile(time_list, 35),3)
        q3 = np.round(np.percentile(time_list, 75),3)
        median = np.round(np.median(time_list),3)
        max = np.round(np.max(time_list),3)
        min = np.round(np.min(time_list),3)

        time_list_smoothed = Plotter.moving_average(time_list, 10)

        plt.style.use("ggplot")
        plt.figure(figsize=(14, 7))

        # Plot des étapes par épisode
        plt.subplot(121)
        plt.plot(time_list, label="Steps per episode", color='blue', alpha=0.6)
        plt.plot(time_list_smoothed, label="Moving average of steps per episode", color='red', linewidth=2)
        plt.legend()
        plt.title("Number of Steps per Episode")
        plt.xlabel("Episode")
        plt.ylabel("Steps")
        plt.grid(True)

        # Boxplot des étapes par épisode
        plt.subplot(122)
        plt.boxplot(time_list, vert=False)
        plt.title(f"Box plot of steps per episode")
        plt.xlabel("Steps")
        plt.grid(True)

        handles, labels = plt.gca().get_legend_handles_labels()
        handles.extend([
            plt.Line2D([0], [0], color='yellow', lw=2),
            plt.Line2D([0], [0], color='yellow', lw=2),
            plt.Line2D([0], [0], color='green', lw=2),
            plt.Line2D([0], [0], color='purple', lw=2),
            plt.Line2D([0], [0], color='red', lw=2),
            plt.Line2D([0], [0], color='blue', lw=2),
            plt.Line2D([0], [0], color='blue', lw=2)
        ])
        labels.extend([
            f'Max : {max} steps', f'Min : {min} steps', 
            f'Mean : {mean} steps', f'Std : {std} steps', 
            f'Q1 : {q1} steps', f'Q3 : {q3} steps', 
            f'Median : {median} steps'
        ])
        
        plt.legend(handles, labels, loc='upper right')

        plt.tight_layout()
        plt.savefig(save_path)

        plt.show()

    @staticmethod
    def plot_multiples(liste, labels, save_path="figure2", title="Multiple plots", xlabel="Episode", ylabel="Value"):
        plt.style.use("ggplot")
        plt.figure(figsize=(14, 7))

        colors = ['blue', 'red', 'green', 'purple', 'orange']
        for i, l in enumerate(liste):
            if save_path == "steps_per_episode_multiple":
                plt.plot(Plotter.moving_average(l, 10), label=labels[i]+" smoothed", linewidth=4, alpha=1, color=colors[i])
            plt.plot(l, label=labels[i], alpha=0.4, color=colors[i])

        plt.legend()
        plt.title(title)
        plt.xlabel(xlabel)
        plt.ylabel(ylabel)
        plt.grid(True)
        plt.tight_layout()
        plt.savefig(save_path)
        plt.show()
    
    @staticmethod
    def compare_plots(results:dict):
        Plotter.plot_multiples(
            [results[i]['lists']["time_list"] for i in range(len(results))], 
            save_path="steps_per_episode_multiple",
            title="Steps per episode",
            xlabel="Episode",
            ylabel="Steps",
            labels=[results[i]["name"] for i in range(len(results))]
        ) 
        Plotter.plot_multiples(
            [results[i]['lists']["mean_reward"] for i in range(len(results))], 
            save_path="reward_per_episode multiple",
            title="Reward per episode",
            xlabel="Episode",
            ylabel="Reward",
            labels=[results[i]["name"] for i in range(len(results))]    
        ) 
        Plotter.plot_multiples(
            [results[i]['lists']["v_list"] for i in range(len(results))], 
            save_path="value_per_episode_multiple",
            title="Norm of value per episode",
            xlabel="Episode",
            ylabel="Norm of value",
            labels=[results[i]["name"] for i in range(len(results))]
        ) 

In [9]:
Plotter.plot_steps_evolution(time_list, "steps_per_episode")
Plotter.plot_steps_evolution(reward_list, "reward_per_episode")
Plotter.plot_steps_evolution(v_list, "value_function_per_episode")

## Actor-critic hyper-parameters

To represent the hyper-parameters of the experiments performed in this notebook, we suggest using the dictionary below.
This dictionary can be read using omegaconf.
Using it is not mandatory.
You can also change the value of hyper-parameters or environment parameters at will.

In [10]:
ac_params = {
    "save_curves": False,
    "save_heatmap": True,
    "mdp": {
        "name": "MazeMDP-v0",
        "width": 5,
        "height": 5,
        "ratio": 0.2,
        "render_mode": "rgb_array",
    },
        
    "log_dir": "./tmp",
    "video_dir": "./tmp/videos",

    "nb_episodes": 100,
    "timeout": 200,
    "render": False, # True, # 
    "nb_repeats": 5,

    "alpha_critic": 0.5,
    "alpha_actor": 0.5,
}

### 3. Test your code

Once everything looks OK, save the obtained plot for your lab report

In [11]:
config = OmegaConf.create(ac_params)

env_config = gym.make(
    config.mdp.name,
    kwargs={
        "width": config.mdp.width,
        "height": config.mdp.height,
        "ratio": config.mdp.ratio,
        "hit": 0.0,
    },
    render_mode=config.mdp.render_mode,
)

a_probs, v, v_list, time_list, reward_list = actor_critic(
    env_config, alpha_critic=config.alpha_critic, 
    alpha_actor=config.alpha_actor, nb_episodes=config.nb_episodes, 
    timeout=config.timeout, render=False
)

/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/gymnasium/core.py:311: UserWarning: WARN: env.gamma to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.gamma` for environment variables or `env.get_wrapper_attr('gamma')` that will search the reminding wrappers.
  logger.warn(


In [12]:
Plotter.plot_steps_evolution(time_list, save_path="steps_per_episode_config_0.png")

# Step 2: Tuning hyper-parameters

In this part, you have to optimize two hyper-parameters of the actor-critic algorithm, namely the actor and critic learning rates.
You have to do so using a simple grid search method and some Bayesian optimization method.
For the latter, we suggest using the default sampler from [optuna](https://optuna.readthedocs.io/en/stable/).
Follow the above link to understand how optuna works.
Note that it also supports grid search and many other hyper-parameters tuning algorithms.

You should make sure that the hyper-parameters tuning algorithms that you compare benefit from the same training budget
We suggest 400 training runs overall for each method,
which means 20 values each for the actor and the critic learning rates in the case of grid search.

## Exercise 2

### 1. Perform hyper-parameters tuning with two algorithms as suggested above.

### 2. Provide a "heatmap" of the norm of the value function given the hyper-parameters, after training for each pair of hyper-parameters.

### 3. Collect the value of the best hyper-parameters found with each algorithm. You will need them for Step 3.

### 4. Include in your report the heatmaps and the best hyper-parameters found for each method.

In [13]:
stats = []

class EnvManager():
    """
    Class to manage the environments
    Methods:
        - generate_envs
    """
    
    @staticmethod
    def generate_envs(nb_env):
        envs = [
            gym.make(
            config.mdp.name,
            kwargs={
                "width": config.mdp.width,
                "height": config.mdp.height,
                "ratio": config.mdp.ratio,
                "hit": 0.0,
            },
            render_mode=config.mdp.render_mode,
            ) for _ in range(nb_env)         
        ]
        return envs
    
envs = EnvManager.generate_envs(5)

class ObjectiveFunction():
    """
    Class to define the objective function to optimize
    Methods:
        - objective_by
    """

    @staticmethod
    def objective_by(trial : optuna.Trial, goal: str = "steps", n_iter: int = 15):
        """
        Args:
            trial: the trial
            goal: the goal to optimize, must be one of {"norm_v, mean_steps", "policy_entropy", "value_stability", "mean_reward"}
        Returns:
            the value of the objective function
        """

    
        global stats

        all_goals = {"norm_v", "mean_steps", "policy_entropy", "value_stability", "mean_reward"}

        if goal not in all_goals:
            raise ValueError(f"Unknown goal: {goal}, must be one of {all_goals}")
        
        alpha_critic = trial.suggest_float("alpha_critic", 1e-2, 1.0)
        alpha_actor = trial.suggest_float("alpha_actor", 1e-2, 1.0)
        

        env_a, env_v, env_v_list, env_time_list, env_mean_reward = [], [], [], [], []
        for env_i in envs:
            all_a, all_v, all_v_list, all_time_list, all_mean_reward = [], [], [], [], []             
            for _ in range(n_iter):
                a_probs, v, v_list, time_list, reward_list = actor_critic(
                    env_i, alpha_critic=alpha_critic, 
                    alpha_actor=alpha_actor, nb_episodes=config.nb_episodes, 
                    timeout=config.timeout, render=False
                )

                mean_reward = np.mean(reward_list)
                all_a.append(a_probs)
                all_v.append(v)
                all_v_list.append(v_list)
                all_time_list.append(np.mean(time_list))
                all_mean_reward.append(mean_reward)

            env_a.append(all_a)
            env_v.append(all_v)
            env_v_list.append(all_v_list)
            env_time_list.append(all_time_list)
            env_mean_reward.append(all_mean_reward)
        
        all_a = np.mean(env_a, axis=0)
        all_v = np.mean(env_v, axis=0)
        all_v_list = np.mean(env_v_list, axis=0)
        all_time_list = np.mean(env_time_list, axis=0)
        all_mean_reward = np.mean(env_mean_reward, axis=0)
        

        time_list_mean = np.mean(all_time_list)
        v_list_mean = np.mean(all_v_list, axis=0)
        a_probs_mean = np.mean(all_a, axis=0)
        v_mean = np.mean(all_v, axis=0)
        mean_reward = np.mean(all_mean_reward)

        stats.append({
            "params" : {
                "alpha_critic": alpha_critic,
                "alpha_actor": alpha_actor,
            },
            "lists" : {
                "time_list": all_time_list,
                "v_list": all_v_list,
                "a_probs": all_a,
                "v": all_v,
                "rewards": all_mean_reward,
            },
            "evaluation_metrics": {
                "mean_steps": -time_list_mean,
                "policy_entropy": -np.sum(a_probs_mean*np.log(a_probs_mean)),
                "value_stability": -np.std(v_list_mean),
                "mean_reward": mean_reward,
                "norm_v": np.max(v_list_mean),
            },
        })

        if goal == "mean_steps":
            return -time_list_mean
        elif goal == "policy_entropy":
            return -np.sum(a_probs_mean*np.log(a_probs_mean))
        elif goal == "value_stability":
            return -np.std(v_list_mean)
        elif goal == "mean_reward":
            return mean_reward
        elif goal == "norm_v":
            return np.max(v_list_mean)

class TuningAlgorithm():
    """
    Class to tune the hyperparameters of the actor-critic algorithm
    Methods:
        - tune_by_bayesian_optimization 
        - tune_by_grid_search
    """

    @staticmethod
    def tune_by_bayesian_optimization(objective = ObjectiveFunction.objective_by, goal: str = "mean_steps"):
        """ 
        Tune the hyperparameters of the actor-critic algorithm using Bayesian optimization
        Args:
            objective: the objective function to optimize
        """
        study = optuna.create_study(direction="maximize")
        objective = partial(objective, goal=goal, n_iter=15)
        study.optimize(objective, n_trials=400)
        
        print(f"Best parameters: {study.best_params}")
        print(f"Best score: {study.best_value}")
        study_results = study.trials_dataframe()
        study_results.to_csv("study_results_bayesian_opt.csv")
        return study.best_params, study 
    
    @staticmethod
    def tune_by_grid_search(objective = ObjectiveFunction.objective_by, goal: str = "mean_steps"):
        """
        Tune the hyperparameters of the actor-critic algorithm using grid search
        Args:
            objective: the objective function to optimize
        """
        search_space = {
            "alpha_critic": np.linspace(1e-2, 1.0, 20),
            "alpha_actor": np.linspace(1e-2, 1.0, 20)
        }
        sampler = optuna.samplers.GridSampler(search_space)
        study = optuna.create_study(direction="maximize", sampler=sampler)
        objective = partial(objective, goal=goal, n_iter=15)
        study.optimize(objective)

        print(f"Best parameters: {study.best_params}")
        print(f"Best score: {study.best_value}")

        study_results = study.trials_dataframe()
        study_results.to_csv("study_results_grid_search.csv")
        return study.best_params, study 

## Tune hyperparameters by grid search

In [21]:
stats = []
TuningAlgorithm.tune_by_grid_search(goal="norm_v")
np.save("stats_grid_search.npy", stats)

[I 2024-09-29 00:35:19,547] A new study created in memory with name: no-name-9972ffaf-4c0e-4ebf-a564-f6828ba319e8
[I 2024-09-29 00:35:22,513] Trial 0 finished with value: 2.7658463265953936 and parameters: {'alpha_critic': 0.6352631578947369, 'alpha_actor': 0.32263157894736844}. Best is trial 0 with value: 2.7658463265953936.
[I 2024-09-29 00:35:25,312] Trial 1 finished with value: 2.7600096476201474 and parameters: {'alpha_critic': 0.4789473684210527, 'alpha_actor': 0.791578947368421}. Best is trial 0 with value: 2.7658463265953936.
[I 2024-09-29 00:35:30,374] Trial 2 finished with value: 1.5115200528470827 and parameters: {'alpha_critic': 0.06210526315789474, 'alpha_actor': 0.8957894736842106}. Best is trial 0 with value: 2.7658463265953936.
[I 2024-09-29 00:35:33,192] Trial 3 finished with value: 2.861439905372243 and parameters: {'alpha_critic': 0.8436842105263158, 'alpha_actor': 0.4789473684210527}. Best is trial 3 with value: 2.861439905372243.
[I 2024-09-29 00:35:36,301] Trial 4

Best parameters: {'alpha_critic': 0.9478947368421053, 'alpha_actor': 0.8957894736842106}
Best score: 2.9141655525059136


## Tune hyperparameters by bayesian optimization

In [22]:
stats = []
TuningAlgorithm.tune_by_bayesian_optimization(goal="norm_v")
np.save("stats_bayesian_optimization.npy", stats)

[I 2024-09-29 01:48:38,691] A new study created in memory with name: no-name-9047f6e3-2207-40ac-80be-f06bb21cf8dc
/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/gymnasium/core.py:311: UserWarning: WARN: env.gamma to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.gamma` for environment variables or `env.get_wrapper_attr('gamma')` that will search the reminding wrappers.
  logger.warn(
[I 2024-09-29 01:48:41,799] Trial 0 finished with value: 2.7031851396711035 and parameters: {'alpha_critic': 0.541068587715524, 'alpha_actor': 0.26478361641097153}. Best is trial 0 with value: 2.7031851396711035.
[I 2024-09-29 01:48:45,333] Trial 1 finished with value: 2.680878993247266 and parameters: {'alpha_critic': 0.8460301305701706, 'alpha_actor': 0.15029122629285344}. Best is trial 0 with value: 2.7031851396711035.
[I 2024-09-29 01:48:48,423] Trial 2 finished with value: 2.5329223331943305 and parameter

Best parameters: {'alpha_critic': 0.8616091421151508, 'alpha_actor': 0.8414187715610656}
Best score: 2.9203500355809733


## Best score Grid Search : 
- ***Best parameters:*** {'alpha_critic': 0.8436842105263158, 'alpha_actor': 1.0}
- ***Best score:*** -7.862479894535781
## Best score Bayesian Optimization  : 
- ***Best parameters:*** {'alpha_critic': 0.804400103698065, 'alpha_actor': 0.9882472809398611}
- ***Best score:*** -7.725128134667201

## GS - value norm 
Best parameters: {'alpha_critic': 0.8957894736842106, 'alpha_actor': 0.9478947368421053}
Best score: 2.995090861297018
## BO - Value norm
Best parameters: {'alpha_critic': 0.8616091421151508, 'alpha_actor': 0.8414187715610656}
Best score: 2.9203500355809733

# Step 3: Statistical tests

Now you have to compare the performance of the actor-critic algorithm tuned
with all the best hyper-parameters you found before, using statistical tests.

The functions below are provided to run Welch's T-test over learning curves.
They have been adapted from a github repository: https://github.com/flowersteam/rl_stats
You don't need to understand them in detail (though it is always a good idea to try to understand more code).

In [14]:
from scipy.stats import ttest_ind
import bootstrapped.bootstrap as bs
import bootstrapped.compare_functions as bs_compare
import bootstrapped.stats_functions as bs_stats

In [15]:
def compute_central_tendency_and_error(id_central, id_error, sample):

    try:
        id_error = int(id_error)
    except:
        pass

    if id_central == "mean":
        central = np.nanmean(sample, axis=1)
    elif id_central == "median":
        central = np.nanmedian(sample, axis=1)
    else:
        raise NotImplementedError

    if isinstance(id_error, int):
        low = np.nanpercentile(sample, q=int((100 - id_error) / 2), axis=1)
        high = np.nanpercentile(sample, q=int(100 - (100 - id_error) / 2), axis=1)
    elif id_error == "std":
        low = central - np.nanstd(sample, axis=1)
        high = central + np.nanstd(sample, axis=1)
    elif id_error == "sem":
        low = central - np.nanstd(sample, axis=1) / np.sqrt(sample.shape[0])
        high = central + np.nanstd(sample, axis=1) / np.sqrt(sample.shape[0])
    else:
        raise NotImplementedError

    return central, low, high

In [16]:
def run_test(test_id, data1, data2, alpha=0.05):
    """
    Compute tests comparing data1 and data2 with confidence level alpha
    :param test_id: (str) refers to what test should be used
    :param data1: (np.ndarray) sample 1
    :param data2: (np.ndarray) sample 2
    :param alpha: (float) confidence level of the test
    :return: (bool) if True, the null hypothesis is rejected
    """
    data1 = data1.squeeze()
    data2 = data2.squeeze()
    n1 = data1.size
    n2 = data2.size

    # perform Welch t-test":
    _, p = ttest_ind(data1, data2, equal_var=False)
    return p < alpha

This last function was adapted for the lab.

In [17]:
def perform_test(perf1, perf2, name1, name2, sample_size=20, downsampling_fact=5, confidence_level=0.01):

    perf1 = perf1.transpose()
    perf2 = perf2.transpose()
    nb_datapoints = perf1.shape[1]
    nb_steps = perf1.shape[0]

    legend = [name1, name2]

    # what do you want to plot ?
    id_central = 'mean' # "median"  # 
    id_error = 80  # (percentiles), also: 'std', 'sem'

    test_id = "Welch t-test"  # recommended
    
    sample1 = perf1[:, np.random.randint(0, nb_datapoints, sample_size)]
    sample2 = perf2[:, np.random.randint(0, nb_datapoints, sample_size)]

    steps = np.arange(0, nb_steps, downsampling_fact)
    sample1 = sample1[steps, :]
    sample2 = sample2[steps, :]

    # test
    sign_diff = np.zeros([len(steps)])
    for i in range(len(steps)):
        sign_diff[i] = run_test(
            test_id, sample1[i, :], sample2[i, :], alpha=confidence_level
        )

    central1, low1, high1 = compute_central_tendency_and_error(
        id_central, id_error, sample1
    )
    central2, low2, high2 = compute_central_tendency_and_error(
        id_central, id_error, sample2
    )

    # plot
    _, ax = plt.subplots(1, 1, figsize=(20, 10))
    lab1 = plt.xlabel("training steps")
    lab2 = plt.ylabel("performance")

    plt.plot(steps, central1, linewidth=7)
    plt.plot(steps, central2, linewidth=7)
    plt.fill_between(steps, low1, high1, alpha=0.3)
    plt.fill_between(steps, low2, high2, alpha=0.3)
    leg = ax.legend(legend, frameon=False)

    # plot significative difference as dots
    idx = np.argwhere(sign_diff == 1)
    y = max(np.nanmax(high1), np.nanmax(high2))
    plt.scatter(steps[idx], y * 1.05 * np.ones([idx.size]), s=100, c="k", marker="o")

    # style
    for line in leg.get_lines():
        line.set_linewidth(10.0)
    ax.spines["top"].set_linewidth(5)
    ax.spines["right"].set_linewidth(5)
    ax.spines["bottom"].set_linewidth(5)
    ax.spines["left"].set_linewidth(5)

    plt.savefig(
        f"./{name1}_{name2}.png", bbox_extra_artists=(leg, lab1, lab2), bbox_inches="tight", dpi=100
    )
    # plt.show()

## Exercise 3

As hyper-parameters, you will use:

- naive tuning, that is a pair (0.5, 0.5) for the actor and critic learning rates,
- the best hyper-parameters you found with the different tuning algorithms you used before.

### 1. For each set of hyper-parameters, collect a large dataset of learning curves.

We suggest using 150 training episodes.

### 2. Perform statistical comparisons

- Take two datasets of learning curves obtained with the hyper-parameters sets that you found with different tuning algorithms.
- Use the ``` perform_test(...)``` function to compare each possible pair of sets.

You should obtain an image for each pair you have tried.
In this image, black dots signal the time step where there is a statistically significant difference between two learning curves.

 ### 3. Conclude.

In [66]:
stats_bayesian_opt_mean_steps = np.load("./results/stats_bayesian_optimization_value_norm.npy", allow_pickle=True)
stats_grid_search_mean_steps = np.load("./results/stats_grid_search_value_norm.npy", allow_pickle=True)

In [67]:
top_k_bayes_opt_mean_steps = sorted(stats_bayesian_opt_mean_steps, key=lambda x: np.mean(x["evaluation_metrics"]["norm_v"]), reverse=True)
top_k_grid_search_mean_steps = sorted(stats_grid_search_mean_steps, key=lambda x: np.mean(x["evaluation_metrics"]["norm_v"]), reverse=True)



print("Bayesian Optimization - Top 1")
print("Metrics: ", top_k_bayes_opt_mean_steps[0]["evaluation_metrics"]) 
print("Params: ", top_k_bayes_opt_mean_steps[0]["params"])

print("Grid Search - Top 1")
print("Metrics: ", top_k_grid_search_mean_steps[0]["evaluation_metrics"])
print("Params: ", top_k_grid_search_mean_steps[0]["params"])


Bayesian Optimization - Top 1
Metrics:  {'mean_steps': -11.718799999999996, 'policy_entropy': 21.820413257660853, 'value_stability': -0.5860787489889543, 'mean_reward': 0.9213333333333334, 'norm_v': 2.9203500355809733}
Params:  {'alpha_critic': 0.8616091421151508, 'alpha_actor': 0.8414187715610656}
Grid Search - Top 1
Metrics:  {'mean_steps': -11.3432, 'policy_entropy': 22.25540670882689, 'value_stability': -0.5248759495698774, 'mean_reward': 0.9290666666666668, 'norm_v': 2.9141655525059136}
Params:  {'alpha_critic': 0.9478947368421053, 'alpha_actor': 0.8957894736842106}


In [78]:
class AnalyseStatistique():
    @staticmethod
    def evaluate_actor_critic(best_params: dict, nb_episodes: int = 150, n_iter=15):
        
        best_alpha_critic = best_params["alpha_critic"]
        best_alpha_actor = best_params["alpha_actor"]
        
        results = {}

        for i, (alpha_critic, alpha_actor) in enumerate(zip(best_alpha_critic, best_alpha_actor)):

            print(f"Running actor-critic with alpha_critic: {alpha_critic}, alpha_actor: {alpha_actor}")
            

            env_v_list, env_time_list, env_mean_reward = [], [], []
            for env_i in envs:
                all_v_list, all_time_list, all_mean_reward = [], [], []
                for _ in range(n_iter):
                    _, _, v_list, time_list, reward_list = actor_critic(
                        env_i, alpha_critic=alpha_critic, alpha_actor=alpha_actor, nb_episodes=nb_episodes, timeout=config.timeout, render=False
                    )
                    all_v_list.append(v_list)
                    all_time_list.append(time_list)
                    all_mean_reward.append(reward_list)

                env_v_list.append(np.mean(all_v_list, axis=0))
                env_time_list.append(np.mean(all_time_list, axis=0))
                env_mean_reward.append(np.mean(all_mean_reward, axis=0))

            
            all_v_list = np.mean(env_v_list, axis=0)
            all_time_list = np.mean(env_time_list, axis=0)
            all_mean_reward = np.mean(env_mean_reward, axis=0)

            results[i] = {
                "statistical_lists": {
                    "all_v_list": np.array(env_v_list),
                    "all_time_list": np.array(env_time_list),
                    "all_mean_reward": np.array(env_mean_reward),
                },
                "lists": {
                    "v_list": all_v_list,
                    "time_list": all_time_list,
                    "mean_reward": all_mean_reward,
                },
                "name" : f"alpha_critic: {np.round(alpha_critic,3)}, alpha_actor: {np.round(alpha_actor,3)}",
                "id": i
            }

        return results
    
    @staticmethod
    def compare_statistical_performance(results:dict):
        names = {
            0: "Default",
            1: "Bayesian Optimization - Value Norm",
            2: "Grid Search - Value Norm",
            3: "Bayesian Optimization - Mean Steps",
            4: "Grid Search - Mean Steps",
        }
        
        for i in range(len(results)-1):
            if i == 0:
                continue
            for j in range(i+1, len(results)):
                if j == 2:
                    continue
                name, name2 = names[i], names[j]
                perform_test(results[i]['statistical_lists']["all_time_list"], results[j]['statistical_lists']["all_time_list"], name, name2+" steps")
                perform_test(results[i]['statistical_lists']["all_mean_reward"], results[j]['statistical_lists']["all_mean_reward"], name, name2+" reward")
                perform_test(results[i]['statistical_lists']["all_v_list"], results[j]['statistical_lists']["all_v_list"], name, name2+" value norm")


In [72]:
best_params = {
    "alpha_critic": [
        0.5, 
        np.float64(top_k_bayes_opt_mean_steps[0]["params"]["alpha_critic"]), 
        np.float64(top_k_grid_search_mean_steps[0]["params"]["alpha_critic"]),
        0.804400103698065,
        0.8436842105263158,
    ],
    "alpha_actor": [
        0.5, 
        np.float64(top_k_bayes_opt_mean_steps[0]["params"]["alpha_actor"]), 
        np.float64(top_k_grid_search_mean_steps[0]["params"]["alpha_actor"]),
        0.9882472809398611,
        1.0,
    ],
}

results = AnalyseStatistique.evaluate_actor_critic(best_params, nb_episodes=150)

Running actor-critic with alpha_critic: 0.5, alpha_actor: 0.5


/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/gymnasium/core.py:311: UserWarning: WARN: env.gamma to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.gamma` for environment variables or `env.get_wrapper_attr('gamma')` that will search the reminding wrappers.
  logger.warn(


Running actor-critic with alpha_critic: 0.8616091421151508, alpha_actor: 0.8414187715610656
Running actor-critic with alpha_critic: 0.9478947368421053, alpha_actor: 0.8957894736842106
Running actor-critic with alpha_critic: 0.804400103698065, alpha_actor: 0.9882472809398611
Running actor-critic with alpha_critic: 0.8436842105263158, alpha_actor: 1.0


In [61]:
Plotter.compare_plots(results)

invalid command name "140358276124160process_stream_events"
    while executing
"140358276124160process_stream_events"
    ("after" script)
can't invoke "event" command: application has been destroyed
    while executing
"event generate $w <<ThemeChanged>>"
    (procedure "ttk::ThemeChanged" line 6)
    invoked from within
"ttk::ThemeChanged"


In [79]:
AnalyseStatistique.compare_statistical_performance(results)

/mnt/c/Cours-Sorbonne/M2/UE_DEEP/.venv/lib/python3.10/site-packages/scipy/stats/_axis_nan_policy.py:573: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  res = hypotest_fun_out(*samples, **kwds)


# Lab report

Your report should contain:
- your source code (probably this notebook), do not forget to put your names on top of the notebook,
- in a separate pdf file with your names in the name of the file:
    + a detailed enough description of the choices you have made: the parameters you have set, the libraries you have used, etc.,
    + the heatmaps obtained using the hyper-parameters tuning algorithms that you have used,
    + the figures resulting from performing Welch's T-test using the best hyper-parameters from the above approaches,
    + your conclusion from these experiments.

Beyond the elements required in this report, any additional studies will be rewarded.
For instance, you can try using a Q-function as critic, using random search as hyper-parameters tuning algorithm,
using more challenging environments, etc.